src/ogma/features/nodeGrouping.ts

```diff
@@ -0,0 +1,205 @@
+import {StyleRule, Transformation, Node, NodeList} from '@linkurious/ogma';
+import {ConflictValue, LkEdgeData, LkNodeData, MissingValue} from '@linkurious/rest-client';
+
+import {LKOgma} from '../index';
+import {Tools} from '../../tools/tools';
+
+export const LKE_NODE_GROUPING_EDGE = 'LKE_NODE_GROUPING_EDGE';
+
+export class NodeGroupingTransformation {
+  public transformation?: Transformation<LkNodeData, LkEdgeData>;
+  public nodeGroupingStyleRule?: StyleRule<LkNodeData, LkEdgeData>;
+  public groupRule?: {ruleName: string; type: string; property: string};
+  private _ogma: LKOgma;
+
+  constructor(ogma: LKOgma) {
+    this._ogma = ogma;
+  }
+
+  /**
+   * Set the grouping rule
+   * @param ruleName The nae of node grouping rule
+   * @param type the type of the node
+   * @param property the property name that will be used to group the nodes
+   */
+  public setGroupingRule(ruleName: string, type: string, property: string): void {
+    this.groupRule = {ruleName: ruleName, type: type, property: property};
+  }
+
+  /**
+   * create a node grouping transformation
+   * It uses groupRule to define the rule
+   * Group the nodes based on a category type and a property value
+   */
+  public async initTransformation(): Promise<void> {
+    if (this.transformation === undefined) {
+      this.transformation = this._ogma.transformations.addNodeGrouping({
+        groupIdFunction: (node) => {
+          if (this._isRuleNotApplicableToNode(node)) {
+            return undefined;
+          } else {
+            const propertyValue = node.getData(['properties', this.groupRule?.property ?? '']);
+            const originalValue =
+              typeof propertyValue === 'object'
+                ? (propertyValue as ConflictValue).original
+                : propertyValue;
+            // groupRule is defined if not we returned undefined
+            return `${this.groupRule!.type}-${originalValue}`.toLowerCase().trim();
+          }
+        },
+        nodeGenerator: (nodes) => {
+          return {
+            data: {
+              categories: [this.groupRule?.type],
```

Comment:
so potentially we return [undefined] here...is this what we want?

Reply:
it will be never undefined. This defines the data of a virtual node and a virtual node will exist only if a node rule is defined.
will change it to "!"